File: src/main/java/ast/Node.java
package ast;

import parser.ParserObject;

import java.util.Optional;

/**
 * A {@link Node} of the abstract syntax tree (AST).
 *
 * @param <C> The {@link Node}s class enumeration.
 * @param <T> The {@link Node}s type enumeration.
 */
public abstract class Node<C extends Enum, T extends Enum> {

    /**
     * The left {@link Node}.
     */
    private Optional<Node<C, T>> left;

    /**
     * The right {@link Node}.
     */
    private Optional<Node<C, T>> right;

    /**
     * The link {@link Node}.
     */
    private Optional<Node<C, T>> link;

    /**
     * The type.
     */
    private C clazz;

    /**
     * The type.
     */
    private T type;

    /**
     * The object.
     */
    private Optional<ParserObject> object;

    /**
     * The constant.
     */
    private Optional<Number> constant;

    /**
     * Create a new {@link Node}.
     */
    public Node() {
        this.left = Optional.empty();
        this.right = Optional.empty();
        this.link = Optional.empty();
        this.object = Optional.empty();
        this.constant = Optional.empty();
    }

    /**
     * Get the left {@link Node}.
     *
     * @return The left {@link Node}.
     */
    public Optional<Node<C, T>> getLeft() {
        return left;
    }

    /**
     * Set the left.
     *
     * @param left The left.
     */
    public void setLeft(Node<C, T> left) {
        this.left = Optional.ofNullable(left);
    }

    /**
     * Get the right {@link Node}.
     *
     * @return The right {@link Node}.
     */
    public Optional<Node<C, T>> getRight() {
        return right;
    }

    /**
     * Set the right {@link Node}.
     *
     * @param right The right {@link Node}.
     */
    public void setRight(Node<C, T> right) {
        this.right = Optional.ofNullable(right);
    }

    /**
     * Get the link {@link Node}.
     *
     * @return The link {@link Node}.
     */
    public Optional<Node<C, T>> getLink() {
        return link;
    }

    /**
     * Set the link {@link Node}.
     *
     * @param link The link {@link Node}.
     */
    public void setLink(Node<C, T> link) {
        this.link = Optional.ofNullable(link);
    }

    /**
     * Get the class.
     *
     * @return The class.
     */
    public C getClazz() {
        return clazz;
    }

    /**
     * Set the class.
     *
     * @param clazz The class.
     */
    public void setClazz(C clazz) {
        this.clazz = clazz;
    }

    /**
     * Get the type.
     *
     * @return The type.
     */
    public T getType() {
        return type;
    }

    /**
     * Set the type.
     *
     * @param type The type.
     */
    public void setType(T type) {
        this.type = type;
    }

    /**
     * Get the object.
     *
     * @return The object.
     */
    public Optional<ParserObject> getObject() {
        return object;
    }

    /**
     * Set the object.
     *
     * @param object The object.
     */
    public void setObject(ParserObject object) {
        this.object = Optional.of(object);
    }

    /**
     * Get the constant.
     *
     * @return The constant.
     */
    public Optional<Number> getConstant() {
        return constant;
    }

    /**
     * Set the constant.
     *
     * @param constant The constant.
     */
    public void setConstant(Number constant) {
        this.constant = Optional.of(constant);
    }

    /**
     * Create the dot representation for this Node.
     *
     * @return The dot representation of this Node.
     */
    public String toDot(final String name) {
        final String cl = clazz == null ? "-" : clazz.toString();
        final String t = type == null ? "-" : type.toString();
        final String o = object.isPresent() ? object.get().getIdentifier() : "-";
        final String co = constant.isPresent() ? constant.get().toString() : "-";

        return "\"" + name + "\" [shape=record, label=\"{" + "<class> " + cl + " | " +
                "<type> " + t + " | " +
                "<object> " + o + " | " +
                "<constant> " + co +
                "}\"];";
    }
}

File: src/main/scala/compiler/ast/Node.scala
package compiler.ast

import java.util.concurrent.{TimeUnit, TimeoutException}

import compiler.parser.ParserObject

import scala.concurrent.duration.Duration
import scala.concurrent.{Await, Future}

/**
  * A [[Node]] of the abstract syntax tree [[Ast]].
  *
  * @param clazz The class.
  * @param typ   The type.
  * @tparam C The [[Node]]s class enumeration.
  * @tparam T The [[Node]]s type enumeration.
  * @author Julian Schlichtholz
  * @version 1.0.0
  */
abstract class Node[C <: Enumeration#Value, T <: Enumeration#Value](var clazz: C = null, var typ: T = null) {

  /**
    * The implementing classes type.
    */
  type Self <: Node[C, T]

  /**
    * The implementations [[compiler.parser.ParserObject]] type.
    */
  type Object <: ParserObject

  /**
    * The constant.
    */
  var constant: Option[Int] = Option.empty

  /**
    * The future [[compiler.parser.ParserObject]]
    */
  var future: Future[Object] = null

  /**
    * The parent [[Node]].
    */
  var parent: Option[Node[C, T]] = Option.empty

  /**
    * The left [[Node]].
    */
  private var _left: Option[Self] = Option.empty

  /**
    * The right [[Node]].
    */
  private var _right: Option[Self] = Option.empty

  /**
    * The link [[Node]].
    */
  private var _link: Option[Self] = Option.empty

  /**
    * The object.
    */
  private var _obj: Option[Object] = Option.empty

  /**
    * Create the dot representation for this Node.
    *
    * @return The dot representation of this Node.
    */
  def toDot(name: String): String = {
    val cl: String = if (clazz == null) "-" else clazz.toString
    val t: String = if (typ == null) "-" else typ.toString
    val o: String = if (obj != null) obj.identifier else "-"
    val co: String = if (constant.isDefined) constant.get.toString else "-"

    "\"" + name + "\" [shape=record, label=\"{" + "<class> " + cl + " | " +
      "<type> " + t + " | " +
      "<object> " + o + " | " +
      "<constant> " + co +
      "}\"];"
  }

  /**
    * Get the [[compiler.parser.ParserObject]].
    *
    * @return The [[compiler.parser.ParserObject]].
    */
  def obj: Object = {
    if (_obj.isDefined) {
      _obj.get
    } else if (future != null) {
      try {
        _obj = Option(Await.result(future, Duration(1000, TimeUnit.MILLISECONDS)))
        _obj.get
      } catch {
        case e: TimeoutException => null.asInstanceOf[Object]
      }
    } else {
      null.asInstanceOf[Object]
    }
  }

  /**
    * Set the [[compiler.parser.ParserObject]].
    *
    * @param parserObject The [[compiler.parser.ParserObject]].
    */
  def obj_=(parserObject: Object): Unit = {
    _obj = Option(parserObject)
  }

  /**
    * Get the left [[Node]].
    *
    * @return The left [[Node]].
    */
  def left: Option[Self] = {
    _left
  }

  /**
    * Set the left [[Node]].
    *
    * @param node The left [[Node]].
    */
  def left_=(node: Self): Unit = {
    if (node != null) node.parent = Option(this)
    _left = Option(node)
  }

  /**
    * Get the right [[Node]].
    *
    * @return The right [[Node]].
    */
  def right: Option[Self] = {
    _right
  }

  /**
    * Set the right [[Node]].
    *
    * @param node The right [[Node]].
    */
  def right_=(node: Self): Unit = {
    node.parent = Option(this.asInstanceOf[Self])
    _right = Option(node)
  }

  /**
    * Get the link [[Node]].
    *
    * @return The link [[Node]].
    */
  def link: Option[Self] = {
    _link
  }

  /**
    * Set the link [[Node]].
    *
    * @param node The link [[Node]].
    */
  def link_=(node: Self): Unit = {
    node.parent = Option(this.asInstanceOf[Self])
    _link = Option(node)
  }
}
